Finish type hinting

scop opened the pull request.

nielstron commented:
Please base your changes on the development branch (`dev`), usually this can be done via `git rebase dev`

nielstron commented on pysyncthru/__init__.py:
```diff
@@ -76,41 +76,41 @@ def is_unknown_state(self) -> bool:
         """Return the model name of the printer."""
         try:
-            return self.data.get("identity").get("model_name")
+            return cast(str, self.data.get("identity", {}).get("model_name"))
```
The cast does not really help anyone but the typchecker and clutters the code unnecessarily.
In addition to cluttering the code, this makes the return type of this method (obscured) to `str` as even the possible `None` returned from `get(...)` is casted to `str`.

Maybe the dictionary can be typed to `Dict[str,str]`?

scop replied:
Sometimes mypy needs a bit of hand holding. Most often fixing its errors lead towards better code though -- such as in this case _if_ the dict can be made more specific. Based on the example in the tests, the `identity` member dict is a `[str,str}, but  I don't think there's a clean way to just mark that as such given the supported Python versions, so a cast or type ignore would be needed anyway.

scop replied:
Added the dict cast -- agree it's a bit better that way. But also even more verbose.

nielstron commented on pysyncthru/tests/test_structure/server_control.py:
```diff
@@ -41,7 +42,7 @@ def stop_server(self):
             self._server.shutdown()
 
-    def start_server(self, timeout=10):
+    def start_server(self, timeout: int = 10) -> None:
```
IMHO the type is already hinted by the default value. However explicitely noting it does not hurt here.

scop replied:
Nothing in the unhinted signature would prevent it from being e.g `Optional` or a `Union` with some other type.

nielstron commented on pysyncthru/tests/test_structure/syncthru_mock_server.py:
```diff
@@ -17,21 +13,21 @@ class SyncThruServer(HTTPServer):
-    def do_GET(self):
-        if self.server.blocked:
+    def do_GET(self) -> None:
+        if cast(SyncThruServer, self.server).blocked:
```
Rather than simply casting to `SyncThruServer` we might overwrite the `__init__` method for type hinting or even throw an error if any server is passed that is not a `SyncThruServer`

scop replied:
`__init__` added.

scop commented:
> Please base your changes on the development branch (`dev`)

If memory serves, I believe I actually did that, and at that time it applied cleanly against both master and dev. I just picked master as the merge target in absence of other information. But sure, can rebase.

scop commented:
Rebased.

Also added dev-packages+travis mypy setup, as there was no replacement install+run for it after 9848b07b5cc36f64a1349ba8d2939b9bc45cb9b7. As mentioned earlier, running mypy the usual way through pre-commit has knkown cross venv issues, and working around them seems a bit much for a project this size (if interested in the details, see how HA works around it).

scop commented:
Hmph, made a bit of a mess with local tests, but things are now.

IMO changing the cast in SyncThruRequestHandler to an overridden __init__ kind of made things worse, SimpleHTTPRequestHandler is not generic so a icky workaround in it to appease the type checker. If it was my call, I'd simply revert back to the cast (or add a type ignore) -- it's in the test suite after all. TBH, I would not have made the test suite code strict-checked at all but would have left it to the default settings, as the S/N ratio of strict checks there is rather low. but for some reason I failed to configure mypy thusly. Well it's done now.

nielstron commented:
Thanks, I think it looks fine the way it is! Thanks for the effort. Configuring mypy for less-strict checking may be left for the future if being problematic
